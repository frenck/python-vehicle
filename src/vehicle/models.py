"""Asynchronous Python client providing RDW vehicle information."""
from __future__ import annotations

from datetime import date, datetime, timezone

try:
    from pydantic.v1 import BaseModel, Field, validator
except ImportError:  # pragma: no cover
    from pydantic import (  # type: ignore[assignment] # pragma: no cover
        BaseModel,
        Field,
        validator,
    )

from .const import (
    VehicleInterior,  # noqa: TCH001
    VehicleOdometerJudgement,  # noqa: TCH001
    VehicleType,  # noqa: TCH001
)


class Vehicle(BaseModel):
    """Object holding vehicle information.

    Attributes
    ----------
        apk_expiration: Expiry date of the APK.
        ascription_date: Date of naming registration of the vehicle.
        ascription_possible: Whether the vehicle is nameable or not.
        brand: Brand of the vehicle.
        energy_label: Energy label of the vehicle.
        engine_capacity: Engine capacity of the vehicle in CC
        exported: Whether the vehicle is exported or not.
        first_admission: First date of registration.
        interior: Interior of the vehicle.
        last_odometer_registration_year: Last year odometer was registered.
        liability_insured: Whether the vehicle is insured or not.
        license_plate: Normalized license plate of the vehicle.
        list_price: List price of the vehicle.
        mass_driveable: Mass of the vehicle when driveable in KG.
        mass_empty: Empty mass of the vehicle in KG.
        model: Model of the vehicle.
        number_of_cylinders: Number of cylinders of the vehicle.
        number_of_doors: Number of doors of the vehicle.
        number_of_seats: Number of seats of the vehicle.
        number_of_wheelchair_seats: Number of wheelchair seats of the vehicle.
        number_of_wheels: Number of wheels of the vehicle.
        odometer_judgement: Odometer judgement of the vehicle.
        pending_recall: Whether the vehicle has a pending recall or not.
        taxi: Whether the vehicle is a taxi or not.
        vehicle_type: Type of the vehicle.
    """

    apk_expiration: date | None = Field(None, alias="vervaldatum_apk")
    ascription_date: date | None = Field(None, alias="datum_tenaamstelling")
    ascription_possible: bool | None = Field(None, alias="tenaamstellen_mogelijk")
    brand: str = Field(..., alias="merk")
    energy_label: str | None = Field(None, alias="zuinigheidslabel")
    engine_capacity: int | None = Field(None, alias="cilinderinhoud")
    exported: bool | None = Field(None, alias="export_indicator")
    interior: VehicleInterior | None = Field(None, alias="inrichting")
    last_odometer_registration_year: int | None = Field(
        None,
        alias="jaar_laatste_registratie_tellerstand",
    )
    liability_insured: bool | None = Field(None, alias="wam_verzekerd")
    license_plate: str = Field(..., alias="kenteken")
    list_price: int | None = Field(None, alias="catalogusprijs")
    first_admission: date | None = Field(None, alias="datum_eerste_toelating")
    mass_empty: int | None = Field(None, alias="massa_ledig_voertuig")
    mass_driveable: int | None = Field(None, alias="massa_rijklaar")
    model: str = Field(..., alias="handelsbenaming")
    number_of_cylinders: int | None = Field(None, alias="aantal_cilinders")
    number_of_doors: int | None = Field(None, alias="aantal_deuren")
    number_of_seats: int | None = Field(None, alias="aantal_zitplaatsen")
    number_of_wheelchair_seats: int | None = Field(
        None,
        alias="aantal_rolstoelplaatsen",
    )
    number_of_wheels: int | None = Field(None, alias="aantal_wielen")
    odometer_judgement: VehicleOdometerJudgement = Field(
        None,
        alias="tellerstandoordeel",
    )
    pending_recall: bool | None = Field(
        None,
        alias="openstaande_terugroepactie_indicator",
    )
    taxi: bool | None = Field(None, alias="taxi_indicator")
    vehicle_type: VehicleType | None = Field(None, alias="voertuigsoort")

    @validator(
        "apk_expiration",
        "ascription_date",
        "first_admission",
        pre=True,
        allow_reuse=True,
    )
    @classmethod
    def parse_date(cls, value: str) -> date:
        """Parse date from string.

        Args:
        ----
            value: String to parse.

        Returns:
        -------
            Parsed date.
        """
        return datetime.strptime(value, "%Y%m%d").replace(tzinfo=timezone.utc).date()

    @validator(
        "ascription_possible",
        "exported",
        "liability_insured",
        "pending_recall",
        "taxi",
        pre=True,
        allow_reuse=True,
    )
    @classmethod
    def parse_bool(cls, value: str) -> bool | None:
        """Parse boolean from string.

        Args:
        ----
            value: String to parse.

        Returns:
        -------
            Parsed boolean.
        """
        return value == "Ja"

    @validator("brand", "model", allow_reuse=True)
    @classmethod
    def make_pretty(cls, value: str) -> str:
        """Parse date from string.

        Args:
        ----
            value: String to make pretty.

        Returns:
        -------
            Pretty string.
        """
        return value.strip().title()

    @validator(
        "interior",
        "odometer_judgement",
        "vehicle_type",
        pre=True,
        allow_reuse=True,
    )
    @classmethod
    def filter_empty(cls, value: str) -> str | None:
        """Filter out empty values.

        Args:
        ----
            value: String to filter.

        Returns:
        -------
            Filtered string.
        """
        if value in {"Niet geregistreerd", "N.v.t."}:
            return None
        return value


Vehicle.update_forward_refs()
