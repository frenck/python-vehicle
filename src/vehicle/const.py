"""Asynchronous Python client providing RDW vehicle information."""

from enum import Enum


class Dataset(str, Enum):
    """Enum holding the RDW dataset identifiers for the Socrata API."""

    PLATED_VEHICLES = "m9d7-ebf2"


class VehicleType(str, Enum):
    """Enum holding the RDW vehicle types."""

    TAILER = "Aanhangwagen"
    PASSENGER_CAR = "Personenauto"
    COMPANY_CAR = "Bedrijfsauto"
    BUS = "Bus"
    MOTORCYCLE = "Motorfiets"
    THREE_WHEELED_MOTOR_VEHICLE = "Driewielig motorrijtuig"
    MOPED = "Bromfiets"
    CENTER_AXLE_TRAILER = "Middenasaanhangwagen"
    AGRICULTURAL_OR_FORESTRY_TRACTOR = "Land- of bosbouwtrekker"
    AGRICULTURAL_OR_FORESTRY_TRACTOR_TRAILER = "Land- of bosb aanhw of getr uitr stuk"


class VehicleInterior(str, Enum):
    """Enum holding the RDW vehicle interior types."""

    CAMPER = "kampeerwagen"
    CARAVAN = "caravan"
    CLOSED_INTERIOR = "gesloten opbouw"
    CONVERTIBLE = "cabriolet"
    COUPE = "coupe"
    FIRETRUCK = "brandweerwagen"
    HATCHBACK = "hatchback"
    LIVESTOCK_TRUCK = "veewagen"
    MPV = "MPV"
    OPEN_LOADING_FLOOR = "open laadvloer"
    OPEN_VEHICLE = "open wagen"
    SEDAN = "sedan"
    STATION_WAGON = "stationwagen"
    PICK_UP_TRUCK = "pick-up truck"


class VehicleOdometerJudgement(str, Enum):
    """Enum holding the RDW vehicle odometer judgement types."""

    NO_JUDGEMENT = "Geen oordeel"
    LOGICAL = "Logisch"
    ILLOGICAL = "Onlogisch"
    UNKNOWN = "UNKNOWN"
