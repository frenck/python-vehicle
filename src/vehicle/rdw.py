"""Asynchronous Python client providing RDW vehicle information."""
from __future__ import annotations

import asyncio
import socket
from dataclasses import dataclass
from importlib import metadata
from typing import Any, Self, cast

import async_timeout
from aiohttp.client import ClientError, ClientResponseError, ClientSession
from aiohttp.hdrs import METH_GET
from yarl import URL

from .const import Dataset
from .exceptions import RDWConnectionError, RDWError, RDWUnknownLicensePlateError
from .models import Vehicle


@dataclass
class RDW:
    """Main class for handling data fetching from RDW."""

    session: ClientSession | None = None
    license_plate: str | None = None
    request_timeout: int = 10
    _close_session: bool = False

    @staticmethod
    def normalize_license_plate(license_plate: str) -> str:
        """Normalize license plate.

        Args:
        ----
            license_plate: License plate.

        Returns:
        -------
            Normalized license plate.
        """
        return license_plate.upper().replace("-", "").replace(" ", "").strip()

    async def _request(
        self,
        dataset: Dataset,
        *,
        data: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Handle a request to a RDW open data (Socrata).

        A generic method for sending/handling HTTP requests done against
        the public RDW data.

        Args:
        ----
            dataset: Identifier for the Socrata dataset to query.
            data: Dictionary of data to send to the Socrata API.

        Returns:
        -------
            A Python dictionary (JSON decoded) with the response from
            the API.

        Raises:
        ------
            RDWConnectionError: An error occurred while communicating with
                the Socrata API.
            RDWError: Received an unexpected response from the Socrata API.
        """
        version = metadata.version(__package__)
        url = URL("https://opendata.rdw.nl/resource/").join(
            URL(f"{dataset.value}.json"),
        )

        headers = {
            "User-Agent": f"PythonVehicle/{version}",
            "Accept": "application/json, text/plain, */*",
        }

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        try:
            async with async_timeout.timeout(self.request_timeout):
                response = await self.session.request(
                    METH_GET,
                    url.with_query(data),
                    headers=headers,
                )
                response.raise_for_status()
        except asyncio.TimeoutError as exception:
            msg = "Timeout occurred while connecting to the Socrata API"
            raise RDWConnectionError(msg) from exception
        except (
            ClientError,
            ClientResponseError,
            socket.gaierror,
        ) as exception:
            msg = "Error occurred while communicating with Socrata API"
            raise RDWConnectionError(msg) from exception

        content_type = response.headers.get("Content-Type", "")
        if "application/json" not in content_type:
            text = await response.text()
            msg = "Unexpected response from the Socrata API"
            raise RDWError(
                msg,
                {"Content-Type": content_type, "response": text},
            )

        return cast(dict[str, Any], await response.json())

    async def vehicle(self, license_plate: str | None = None) -> Vehicle:
        """Get devices information about a Vehicle.

        Args:
        ----
            license_plate: License plate of the vehicle, if not provided
                the license plate of the object will be used.

        Returns:
        -------
            A Vehicle object, with information about the vehicle.

        Raises:
        ------
            RDWError: No license plate provided.
            RDWUnknownLicensePlateError: License plate not found in RDW Socrata DB.
        """
        license_plate = license_plate or self.license_plate
        if license_plate is None:
            msg = "No license plate provided"
            raise RDWError(msg)

        data = await self._request(
            Dataset.PLATED_VEHICLES,
            data={"kenteken": self.normalize_license_plate(license_plate)},
        )
        if not data:
            msg = f"License plate {license_plate} not found in RDW Socrata database"
            raise RDWUnknownLicensePlateError(msg)

        return Vehicle.parse_obj(data[0])  # type: ignore[index]

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> Self:
        """Async enter.

        Returns
        -------
            The RDW object.
        """
        return self

    async def __aexit__(self, *_exc_info: object) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.
        """
        await self.close()
