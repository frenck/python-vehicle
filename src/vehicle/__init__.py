"""Asynchronous Python client providing RDW vehicle information."""
from .const import VehicleInterior, VehicleOdometerJudgement, VehicleType
from .exceptions import RDWConnectionError, RDWError, RDWUnknownLicensePlateError
from .models import Vehicle
from .rdw import RDW

__all__ = [
    "RDW",
    "RDWConnectionError",
    "RDWError",
    "RDWUnknownLicensePlateError",
    "Vehicle",
    "VehicleInterior",
    "VehicleOdometerJudgement",
    "VehicleType",
]
