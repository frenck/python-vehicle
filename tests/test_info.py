"""Tests for the vehicle Library."""
from datetime import date

import aiohttp
import pytest

from vehicle import (
    RDW,
    RDWError,
    RDWUnknownLicensePlateError,
    Vehicle,
    VehicleInterior,
    VehicleOdometerJudgement,
    VehicleType,
)
from vehicle.const import Dataset

from . import load_fixture


@pytest.mark.asyncio
async def test_vehicle_data(aresponses):  # pylint: disable=too-many-statements
    """Test getting Vehicle information."""
    aresponses.add(
        "opendata.rdw.nl",
        f"/resource/{Dataset.PLATED_VEHICLES}.json",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("11ZKZ3.json"),
        ),
    )
    async with aiohttp.ClientSession() as session:
        rdw = RDW(session=session)
        vehicle: Vehicle = await rdw.vehicle("11-ZKZ-3")
        assert vehicle
        assert vehicle.apk_expiration == date(2022, 1, 4)
        assert vehicle.ascription_date == date(2021, 11, 4)
        assert vehicle.ascription_possible is True
        assert vehicle.brand == "Skoda"
        assert vehicle.energy_label == "A"
        assert vehicle.engine_capacity == 999
        assert vehicle.exported is False
        assert vehicle.first_admission == date(2013, 1, 4)
        assert vehicle.first_admission_netherlands == date(2013, 1, 4)
        assert vehicle.interior == VehicleInterior.HATCHBACK
        assert vehicle.last_odometer_registration_year == 2021
        assert vehicle.liability_insured is False
        assert vehicle.license_plate == "11ZKZ3"
        assert vehicle.list_price == 10697
        assert vehicle.mass_driveable == 940
        assert vehicle.mass_empty == 840
        assert vehicle.model == "Citigo"
        assert vehicle.number_of_cylinders == 3
        assert vehicle.number_of_doors == 0
        assert vehicle.number_of_seats == 4
        assert vehicle.number_of_wheelchair_seats == 0
        assert vehicle.number_of_wheels == 4
        assert vehicle.odometer_judgement == VehicleOdometerJudgement.LOGICAL
        assert vehicle.pending_recall is False
        assert vehicle.taxi is None
        assert vehicle.vehicle_type == VehicleType.PASSENGER_CAR

    aresponses.add(
        "opendata.rdw.nl",
        f"/resource/{Dataset.PLATED_VEHICLES}.json",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("0001TJ.json"),
        ),
    )
    async with aiohttp.ClientSession() as session:
        rdw = RDW(session=session)
        vehicle: Vehicle = await rdw.vehicle("00-01-TJ")
        assert vehicle
        assert vehicle.apk_expiration == date(2023, 7, 26)
        assert vehicle.ascription_date == date(2013, 7, 25)
        assert vehicle.ascription_possible is True
        assert vehicle.brand == "Ford"
        assert vehicle.energy_label is None
        assert vehicle.engine_capacity is None
        assert vehicle.exported is False
        assert vehicle.first_admission == date(1972, 1, 13)
        assert vehicle.first_admission_netherlands == date(1972, 1, 13)
        assert vehicle.interior is None
        assert vehicle.last_odometer_registration_year == 2021
        assert vehicle.liability_insured is True
        assert vehicle.license_plate == "0001TJ"
        assert vehicle.list_price is None
        assert vehicle.mass_driveable == 950
        assert vehicle.mass_empty == 850
        assert vehicle.model == "Escort Mexico"
        assert vehicle.number_of_cylinders == 4
        assert vehicle.number_of_doors == 2
        assert vehicle.number_of_seats is None
        assert vehicle.number_of_wheelchair_seats is None
        assert vehicle.number_of_wheels == 4
        assert vehicle.odometer_judgement == VehicleOdometerJudgement.NO_JUDGEMENT
        assert vehicle.pending_recall is False
        assert vehicle.taxi is False
        assert vehicle.vehicle_type == VehicleType.PASSENGER_CAR


@pytest.mark.asyncio
async def test_no_vehicle(aresponses):
    """Test getting non-existing Vehicle."""
    aresponses.add(
        "opendata.rdw.nl",
        f"/resource/{Dataset.PLATED_VEHICLES}.json",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("no_vehicles.json"),
        ),
    )
    async with aiohttp.ClientSession() as session:
        rdw = RDW(session=session)
        with pytest.raises(RDWUnknownLicensePlateError):
            await rdw.vehicle("00-00-00")


@pytest.mark.asyncio
async def test_no_license_plate_provided():
    """Test getting Vehicle without license plate."""
    rdw = RDW()
    with pytest.raises(RDWError):
        await rdw.vehicle()
