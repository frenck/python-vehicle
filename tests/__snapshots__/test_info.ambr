# serializer version: 1
# name: test_vehicle_data[0001TJ]
  Vehicle(brand='Ford', license_plate='0001TJ', model='Escort Mexico', apk_expiration=datetime.date(2023, 7, 26), ascription_date=datetime.date(2013, 7, 25), ascription_possible=True, energy_label=None, engine_capacity=None, exported=False, interior=None, last_odometer_registration_year=2021, liability_insured=True, list_price=None, first_admission=datetime.date(1972, 1, 13), mass_empty=850, mass_driveable=950, number_of_cylinders=4, number_of_doors=2, number_of_seats=None, number_of_wheelchair_seats=None, number_of_wheels=4, odometer_judgement=<VehicleOdometerJudgement.NO_JUDGEMENT: 'Geen oordeel'>, pending_recall=False, taxi=False, vehicle_type=<VehicleType.PASSENGER_CAR: 'Personenauto'>)
# ---
# name: test_vehicle_data[11ZKZ3]
  Vehicle(brand='Skoda', license_plate='11ZKZ3', model='Citigo', apk_expiration=datetime.date(2022, 1, 4), ascription_date=datetime.date(2021, 11, 4), ascription_possible=True, energy_label='A', engine_capacity=999, exported=False, interior=<VehicleInterior.HATCHBACK: 'hatchback'>, last_odometer_registration_year=2021, liability_insured=False, list_price=10697, first_admission=datetime.date(2013, 1, 4), mass_empty=840, mass_driveable=940, number_of_cylinders=3, number_of_doors=0, number_of_seats=4, number_of_wheelchair_seats=0, number_of_wheels=4, odometer_judgement=<VehicleOdometerJudgement.LOGICAL: 'Logisch'>, pending_recall=False, taxi=None, vehicle_type=<VehicleType.PASSENGER_CAR: 'Personenauto'>)
# ---
# name: test_vehicle_data[VXJ99N]
  Vehicle(brand='Opel', license_plate='VXJ99N', model='Vivaro', apk_expiration=None, ascription_date=None, ascription_possible=True, energy_label=None, engine_capacity=1997, exported=False, interior=<VehicleInterior.CLOSED_INTERIOR: 'gesloten opbouw'>, last_odometer_registration_year=None, liability_insured=False, list_price=55421, first_admission=None, mass_empty=1745, mass_driveable=1845, number_of_cylinders=4, number_of_doors=2, number_of_seats=6, number_of_wheelchair_seats=None, number_of_wheels=4, odometer_judgement=None, pending_recall=False, taxi=False, vehicle_type=<VehicleType.COMPANY_CAR: 'Bedrijfsauto'>)
# ---
